File: web/inspector/src/lib/fixtures.ts
/**
 * fixtures.ts — the inspector's fixture switcher.
 *
 * The fixture switcher (RFC §12) lets a developer exercise an App's UI states
 * — happy / empty / error / permission / slow / large — WITHOUT a running
 * backend. Each fixture is derived FROM the tool's generated output contract
 * (`contracts.ts`): the inspector owns the contracts (P1 — contract-first), so
 * a fixture is always a structurally valid instance of the schema it stands
 * for, never hand-written drift.
 *
 * Selecting a fixture feeds the App synthetic `structuredContent` of the
 * chosen shape via the host-half bridge's `sendToolResult` — closing Phase
 * 22's `tools/call` not-wired seam: a `tools/call` in the inspector is
 * answered from the active fixture.
 */

import type { ContractSchema, ToolContract } from './contracts.js';

/** The six fixture kinds the switcher offers (RFC §12). */
export const FIXTURE_KINDS = [
  'happy',
  'empty',
  'error',
  'permission',
  'slow',
  'large',
] as const;

/** One of the six fixture kinds. */
export type FixtureKind = (typeof FIXTURE_KINDS)[number];

/** A fixture's human label + one-line description, for the switcher UI. */
export const FIXTURE_META: Record<
  FixtureKind,
  { label: string; description: string }
> = {
  happy: {
    label: 'Happy path',
    description: 'A fully populated, valid result — the App renders its ready state.',
  },
  empty: {
    label: 'Empty',
    description: 'A structurally valid but empty result — the App renders its empty state.',
  },
  error: {
    label: 'Error',
    description: 'A tool error — the App renders its error state with retry.',
  },
  permission: {
    label: 'Permission denied',
    description: 'A permission-denied error — the App renders its permission state.',
  },
  slow: {
    label: 'Slow',
    description: 'A delayed result — the App renders its loading state before resolving.',
  },
  large: {
    label: 'Large',
    description: 'A high-volume result — the App renders under a large dataset.',
  },
};

/** The synthetic `tools/call` outcome a fixture feeds the App. */
export interface Fixture {
  /** Which of the six kinds this is. */
  kind: FixtureKind;
  /** True when the fixture stands for a failed tool call. */
  isError: boolean;
  /** The synthetic `structuredContent` — present for a successful fixture. */
  structuredContent?: unknown;
  /** A plain-text content line, mirroring a tool Result's `Text`. */
  text: string;
  /** An error message — present when `isError`. */
  error?: { code: number; message: string };
  /** An artificial delay (ms) before the result resolves — the slow fixture. */
  delayMs: number;
}

/** The artificial delay the `slow` fixture injects before resolving. */
export const SLOW_FIXTURE_DELAY_MS = 1500;

/** The row count the `large` fixture generates for an array-typed field. */
export const LARGE_FIXTURE_ROWS = 250;

/**
 * A project-on-disk fixture loaded from `fixtures/<tool>/<kind>.json` (Phase
 * 24, decision D-126). The inspector backend's `/api/fixtures` endpoint
 * serves these; when present, [buildFixtures] uses them in preference to the
 * schema-derived synthetic payloads so the App renders realistic data
 * (real chart values, real table rows, a real metric card) rather than the
 * `"sample-value"` placeholders the schema instantiator can produce.
 */
export interface ProjectFixture {
  tool: string;
  kind: FixtureKind;
  description?: string;
  state: string;
  input?: Record<string, unknown>;
  structuredContent?: Record<string, unknown>;
}

/**
 * Builds the six fixtures for a tool contract. When a project on-disk fixture
 * is provided in `overrides`, its `structuredContent` (and description) take
 * precedence over the schema-derived synthetic payload — the inspector's
 * /api/fixtures loader (Phase 24) wires this. With no override, every fixture
 * is generated from `contract.outputSchema` so it remains a structurally
 * valid instance of the tool's generated output contract (P1).
 */
export function buildFixtures(
  contract: ToolContract,
  overrides?: Partial<Record<FixtureKind, ProjectFixture>>,
): Record<FixtureKind, Fixture> {
  const schema = contract.outputSchema;
  const pick = (kind: FixtureKind, fallback: unknown): unknown => {
    const o = overrides?.[kind];
    if (o?.structuredContent && Object.keys(o.structuredContent).length > 0) {
      return o.structuredContent;
    }
    return fallback;
  };
  return {
    happy: {
      kind: 'happy',
      isError: false,
      structuredContent: pick('happy', instantiate(schema, 'happy')),
      text: `${contract.name}: result ready`,
      delayMs: 0,
    },
    empty: {
      kind: 'empty',
      isError: false,
      structuredContent: pick('empty', instantiate(schema, 'empty')),
      text: `${contract.name}: no results`,
      delayMs: 0,
    },
    error: {
      kind: 'error',
      isError: true,
      structuredContent: overrides?.error?.structuredContent,
      text: `${contract.name}: tool error`,
      error: { code: -32000, message: 'the tool handler returned an error' },
      delayMs: 0,
    },
    permission: {
      kind: 'permission',
      isError: true,
      structuredContent: overrides?.permission?.structuredContent,
      text: `${contract.name}: permission denied`,
      error: { code: -32003, message: 'permission denied for this tool' },
      delayMs: 0,
    },
    slow: {
      kind: 'slow',
      isError: false,
      structuredContent: pick('slow', instantiate(schema, 'happy')),
      text: `${contract.name}: result ready (delayed)`,
      delayMs: SLOW_FIXTURE_DELAY_MS,
    },
    large: {
      kind: 'large',
      isError: false,
      structuredContent: pick('large', instantiate(schema, 'large')),
      text: `${contract.name}: ${LARGE_FIXTURE_ROWS} results`,
      delayMs: 0,
    },
  };
}

/** Fixture-generation mode, varying the volume an array-typed field carries. */
type Mode = 'happy' | 'empty' | 'large';

/**
 * Instantiates a JSON-Schema as a concrete value. It is a pure, total mapping
 * — every schema yields a value, an unknown schema yields null — so a fixture
 * is always producible from any generated contract.
 */
export function instantiate(
  schema: ContractSchema | undefined,
  mode: Mode,
): unknown {
  if (!schema) return null;
  switch (schema.type) {
    case 'object':
      return instantiateObject(schema, mode);
    case 'array':
      return instantiateArray(schema, mode);
    case 'string':
      return instantiateString(schema, mode);
    case 'number':
    case 'integer':
      return mode === 'empty' ? 0 : sampleNumber(mode);
    case 'boolean':
      return mode !== 'empty';
    default:
      // An unconstrained schema (`true` — e.g. json.RawMessage) or an
      // unrecognised type: a tolerant null, never a throw.
      return schema.enum && schema.enum.length > 0 ? schema.enum[0] : null;
  }
}

function instantiateObject(
  schema: ContractSchema,
  mode: Mode,
): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  const props = schema.properties ?? {};
  for (const [key, propSchema] of Object.entries(props)) {
    out[key] = instantiate(propSchema, mode);
  }
  return out;
}

function instantiateArray(schema: ContractSchema, mode: Mode): unknown[] {
  if (mode === 'empty') return [];
  const count = mode === 'large' ? LARGE_FIXTURE_ROWS : 3;
  const item = schema.items;
  const out: unknown[] = [];
  for (let i = 0; i < count; i++) {
    out.push(instantiate(item, mode === 'large' ? 'large' : 'happy'));
  }
  return out;
}

function instantiateString(schema: ContractSchema, mode: Mode): string {
  if (schema.enum && schema.enum.length > 0) {
    return String(schema.enum[0]);
  }
  if (mode === 'empty') return '';
  if (schema.format === 'date-time') {
    return new Date(0).toISOString();
  }
  return mode === 'large' ? 'sample-value-large' : 'sample-value';
}

function sampleNumber(mode: Mode): number {
  return mode === 'large' ? 999999 : 42;
}

File: internal/inspector/fixtures.go
// Package inspector — on-disk fixture loader (Phase 24, D-126).
//
// The Phase 23 fixture switcher synthesises structuredContent from the tool's
// generated output schema. That is structurally correct (P1) but the synthesised
// values are placeholders — "sample-value" strings, 42 numbers — so the App's
// dispatcher does not see a kind it knows ("chart" / "table" / "metric_card")
// and the rendered widget can never be the realistic data the template ships.
//
// The fixture-loader below complements the switcher: when a project on disk
// carries a `fixtures/<tool>/<kind>.json` tree (the analytics-widgets template
// ships eighteen), the inspector loads them, and the frontend switcher prefers
// the on-disk payload over the synthesised one. The schema-derived path
// remains the fallback for projects that ship no fixtures, so the existing
// behaviour is preserved.
//
// The loader is read-only. It performs no `tools/call`; it only reads files
// the developer's own project carries. RFC §12 / P4 (the inspector is a
// dev surface, never a production MCP client) is preserved.
package inspector

import (
	"encoding/json"
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// FixtureKind is one of the six fixture kinds (RFC §12, brief 04 §2.2).
type FixtureKind string

// The six fixture kinds the switcher offers.
const (
	FixtureHappy      FixtureKind = "happy"
	FixtureEmpty      FixtureKind = "empty"
	FixtureError      FixtureKind = "error"
	FixturePermission FixtureKind = "permission"
	FixtureSlow       FixtureKind = "slow"
	FixtureLarge      FixtureKind = "large"
)

// fixtureKinds lists the kinds in a stable order — the loader scans them in
// that order so the inspector picker is deterministic.
var fixtureKinds = []FixtureKind{
	FixtureHappy,
	FixtureEmpty,
	FixtureError,
	FixturePermission,
	FixtureSlow,
	FixtureLarge,
}

// rawFixture is the on-disk JSON shape the analytics-widgets template uses
// (and any future template ships). It carries the tool input the model would
// pass, the UI state the fixture represents, and an optional output_override
// the App's dispatcher renders directly. Unknown fields are tolerated.
type rawFixture struct {
	Name           string         `json:"name"`
	Description    string         `json:"description"`
	State          string         `json:"state"`
	Input          map[string]any `json:"input"`
	OutputOverride map[string]any `json:"output_override,omitempty"`
}

// ProjectFixture is one fixture surfaced to the frontend switcher. It is the
// inspector's own type — no JSON wire shape from a template leaks through
// unchanged (the loader normalises every field).
type ProjectFixture struct {
	// Tool is the tool name this fixture belongs to (e.g. "create_chart").
	Tool string `json:"tool"`
	// Kind is one of the six FixtureKind values.
	Kind FixtureKind `json:"kind"`
	// Description is the human-readable summary from the JSON.
	Description string `json:"description,omitempty"`
	// State is the UI state the fixture stands for ("ready"/"empty"/...).
	State string `json:"state"`
	// Input is the model-side input — surfaced for the inspector's RPC log
	// and the integration test, never used by the App's renderer.
	Input map[string]any `json:"input,omitempty"`
	// StructuredContent is the payload the App's dispatcher receives. When
	// the on-disk fixture carries `output_override`, it is used verbatim;
	// otherwise the loader derives it from `input` by pinning `kind` and
	// `state` so the App's discriminator routes correctly.
	StructuredContent map[string]any `json:"structuredContent"`
}

// FixtureSource produces the project's on-disk fixtures, by tool. It is
// invoked per `GET /api/fixtures` request — the inspector re-reads on every
// request so a developer can edit a fixture without restarting.
type FixtureSource func() ([]ProjectFixture, error)

// FixturesFromDir builds a FixtureSource that reads from `<dir>/fixtures/`.
// An empty dir, or a missing fixtures/ subtree, returns an empty slice and a
// nil error — the inspector degrades to the schema-derived synthetic
// fixtures. A malformed JSON file is a typed error so the developer notices.
func FixturesFromDir(dir string) FixtureSource {
	return func() ([]ProjectFixture, error) {
		if strings.TrimSpace(dir) == "" {
			return []ProjectFixture{}, nil
		}
		root := filepath.Join(dir, "fixtures")
		info, err := os.Stat(root)
		if err != nil {
			if errors.Is(err, fs.ErrNotExist) {
				return []ProjectFixture{}, nil
			}
			return nil, fmt.Errorf("dockyard/internal/inspector: stat %s: %w", root, err)
		}
		if !info.IsDir() {
			return []ProjectFixture{}, nil
		}
		return readFixtureTree(root)
	}
}

// readFixtureTree walks the fixtures/ directory and loads every `<tool>/<kind>.json`
// it finds. Only the six known kinds are loaded; an unrecognised filename is
// skipped (forward-compat: a template may ship more under a different
// switcher in the future). The returned slice is sorted by tool then by the
// fixed kind order so the frontend picker is deterministic.
func readFixtureTree(root string) ([]ProjectFixture, error) {
	tools, err := os.ReadDir(root)
	if err != nil {
		return nil, fmt.Errorf("dockyard/internal/inspector: read %s: %w", root, err)
	}
	out := make([]ProjectFixture, 0, len(tools)*len(fixtureKinds))
	for _, t := range tools {
		if !t.IsDir() {
			continue
		}
		tool := t.Name()
		toolDir := filepath.Join(root, tool)
		for _, kind := range fixtureKinds {
			path := filepath.Join(toolDir, string(kind)+".json")
			info, statErr := os.Stat(path)
			if statErr != nil || info.IsDir() {
				continue
			}
			fixture, loadErr := loadFixture(tool, kind, path)
			if loadErr != nil {
				return nil, loadErr
			}
			out = append(out, fixture)
		}
	}
	sort.Slice(out, func(i, j int) bool {
		if out[i].Tool != out[j].Tool {
			return out[i].Tool < out[j].Tool
		}
		return kindOrder(out[i].Kind) < kindOrder(out[j].Kind)
	})
	return out, nil
}

// loadFixture reads one fixture JSON, normalises it into a ProjectFixture,
// and derives a structuredContent payload when the JSON ships no
// `output_override`. The `path` is constructed from a controlled
// `fixtures/<tool>/<kind>.json` shape where `tool` comes from a
// `ReadDir` of the project's fixtures/ subtree and `kind` is one of the
// six well-known FixtureKind values — there is no user-supplied input on
// this path. The inspector itself is dev-mode-gated and localhost-only
// (CLAUDE.md §7).
func loadFixture(tool string, kind FixtureKind, path string) (ProjectFixture, error) {
	body, err := os.ReadFile(path) //nolint:gosec // G304: path is bounded to fixtures/<tool>/<kind>.json by the walker above.
	if err != nil {
		return ProjectFixture{}, fmt.Errorf("dockyard/internal/inspector: read %s: %w", path, err)
	}
	var raw rawFixture
	if err := json.Unmarshal(body, &raw); err != nil {
		return ProjectFixture{}, fmt.Errorf("dockyard/internal/inspector: parse %s: %w", path, err)
	}
	state := strings.TrimSpace(raw.State)
	if state == "" {
		state = string(kind)
	}
	sc := raw.OutputOverride
	if len(sc) == 0 {
		sc = deriveStructuredContent(tool, state, raw.Input)
	}
	return ProjectFixture{
		Tool:              tool,
		Kind:              kind,
		Description:       raw.Description,
		State:             state,
		Input:             raw.Input,
		StructuredContent: sc,
	}, nil
}

// deriveStructuredContent builds a structuredContent payload from a fixture's
// input by pinning the well-known dispatcher fields (`kind`, `state`,
// `theme`). This mirrors the analytics-widgets handlers: every handler
// constructs its output by copying its input and adding `kind` + `state`.
// Templates that need richer derivation should ship `output_override`.
//
// The toolToKind mapping is intentionally explicit — RFC §6 / brief 04: the
// inspector knows the V1 template set's dispatcher discriminators, and any
// future template either ships an `output_override` (recommended) or chooses
// one of the known kinds. An unrecognised tool name produces a `state`-only
// payload, which the App will route to its loading/unknown branch — a clean,
// loud fallback the developer can spot rather than a silent placeholder.
func deriveStructuredContent(tool, state string, input map[string]any) map[string]any {
	out := map[string]any{}
	for k, v := range input {
		out[k] = v
	}
	if kind, ok := toolToKind[tool]; ok {
		out["kind"] = kind
	}
	out["state"] = state
	if _, hasTheme := out["theme"]; !hasTheme {
		out["theme"] = "auto"
	}
	return out
}

// toolToKind is the discriminator mapping for the V1 template set. The
// analytics-widgets template ships exactly these three tools (Phase 24).
var toolToKind = map[string]string{
	"create_chart":       "chart",
	"create_table":       "table",
	"create_metric_card": "metric_card",
}

// kindOrder returns the stable index of a FixtureKind for sorting.
func kindOrder(k FixtureKind) int {
	for i, kk := range fixtureKinds {
		if kk == k {
			return i
		}
	}
	return len(fixtureKinds)
}
